src/adapter.rs

```diff
@@ -113,10 +124,140 @@ pub enum TokenKind<'a> {
+    arguments: Option<Vec<AttributeValue>>,
+}
+
+impl<'a> TryFrom<&'a str> for AttributeValue {
+    type Error = anyhow::Error;
+
+    // TODO: Are we sure it recognises all possible attributes? Maybe it's better
+    // to ignore errors instead of panicing?
+    fn try_from(as_string: &'a str) -> anyhow::Result<Self> {
+        const PATH_RE_STR: &str = r"[[:blank:]]*(?P<simple_path>[[:word:]:]+)[[:blank:]]*";
+        lazy_static! {
+            static ref PATH_RE: Regex = Regex::new(format!(r"^{}$", PATH_RE_STR).as_str()).unwrap();
+            static ref ASSIGNMENT_RE: Regex =
+                Regex::new(format!(r"^{}=[[:blank:]]*(?P<expression>.*)$", PATH_RE_STR).as_str())
+                    .unwrap();
+            static ref ARGUMENTS_RE: Regex = Regex::new(
+                format!(
+                    r"^{}[\(\[\{{](?P<arguments>.*)[\)\]\}}][[:blank:]]*$",
+                    PATH_RE_STR
+                )
+                .as_str()
+            )
+            .unwrap();
+        }
+
+        PATH_RE
+            .captures(as_string)
+            .map(|captures| AttributeValue {
+                as_string: as_string.to_string(),
+                base: captures["simple_path"].to_string(),
+                assigned_expression: None,
+                arguments: None,
+            })
+            .or_else(|| {
+                ASSIGNMENT_RE
+                    .captures(as_string)
+                    .map(|captures| AttributeValue {
+                        as_string: as_string.to_string(),
+                        base: captures["simple_path"].to_string(),
+                        assigned_expression: Some(captures["expression"].to_string()),
+                        arguments: None,
+                    })
+            })
+            .or_else(|| {
+                ARGUMENTS_RE.captures(as_string).map(|captures| {
+                    let mut i = 0;
+                    let mut depth = 0;
+                    let ref arg_str = captures["arguments"];
+                    let mut arguments: Vec<AttributeValue> = Vec::new();
+                    let mut only_white = true;
+                    for (j, c) in arg_str.chars().enumerate() {
+                        if c != ' ' && c != '\t' {
+                            only_white = false;
+                        }
+
+                        if c == '(' || c == '[' || c == '{' {
+                            depth += 1;
+                        } else if c == ')' || c == ']' || c == '}' {
+                            depth -= 1;
+                        } else if c == ',' && depth == 0 {
+                            arguments.push(AttributeValue::try_from(&arg_str[i..j]).unwrap());
+                            i = j + 1;
+                            only_white = true;
+                        }
+                    }
+                    if i < arg_str.len() && !only_white {
+                        arguments.push(AttributeValue::try_from(&arg_str[i..]).unwrap());
+                    }
```

Comment:
I know it's very easy to slip into "IOI mode," but it's very important to not do that — especially in code that is going to be long-lived, and will be maintained for a long time by people who aren't just yourselves. This code is practically unreadable to anyone that didn't do IOI / ACM and hasn't been trained on understanding it quickly for purposes of submitting challenges on TopCoder and similar sites :)

Things you can do to improve this:
- Use more descriptive variable names than `i, j, c` etc.
- Write comments describing what is going on.
- Extract this code out of `TryFrom` and even out of `adapter.rs`, into its own file and break it up into separate, smaller functions that are well-named, individually-tested, and documented.

Reply:
You're right, probably even me after a year wouldn't be able to understand this code easily... I'll improve it!

What about making a separate module `attributes`? Or maybe it should be a sub-module of adapter? Definitions of Attribute and AttributeValue, parsing them and specific tests would be more then two hundred lines of code, which in fact does not depend on anything else from the crate.